File: code/acute_pilot/penalty_sa/sa.py
#!/usr/bin/env python3
"""Fixed-cardinality penalty SA for OEIS A089676 acute sets in {0,1}^n.

Energy = number of ordered (apex Q; legs P,R) right-angle violations:
  (P^Q)&(R^Q)==0, over all ordered triples of the current size-k array S.
We hold k FIXED and drive energy->0 by replacing one vertex at a time
(simulated annealing). energy==0  =>  size-k acute set.

Right-angle count is symmetric in the two legs, so we count UNORDERED leg
pairs and the verifier (which counts ordered) agrees on the zero set: a set
has 0 unordered violations iff 0 ordered violations.

Usage: sa.py <n> <k> [seeds] [iters] [seedfile] [out_prefix]
Any energy-0 result is written and MUST be checked with verify.py.
"""
import sys, os, random, re

def parse_witnesses(path):
    txt=open(path).read(); W={}
    parts=re.split(r'a\((\d+)\)\s*>=\s*(\d+):', txt)
    for i in range(1,len(parts),3):
        nn=int(parts[i])
        rows=re.findall(r'\(([01\s]+)\)', parts[i+2])
        masks=[]
        for r in rows:
            bits=[c for c in r if c in '01']
            v=0
            for j,c in enumerate(bits):
                if c=='1': v|=(1<<j)
            masks.append(v)
        W[nn]=masks
    return W

def right_pair(a,b,q,pc):
    # is there a right angle at apex q between legs a,b ?  ((a^q)&(b^q))==0
    return ((a^q)&(b^q))==0

def total_energy(S):
    m=len(S); e=0
    for j in range(m):
        q=S[j]
        for a in range(m):
            if a==j: continue
            xa=S[a]^q
            for b in range(a+1,m):
                if b==j: continue
                if (xa&(S[b]^q))==0:
                    e+=1
    return e

def vertex_contrib(S, idx, val):
    """Total violations that INVOLVE position idx (with S[idx]=val), counted once each.
    Covers: idx as apex; idx as one leg (any apex, other leg). Used for delta moves."""
    m=len(S); e=0
    # idx as apex
    for a in range(m):
        if a==idx: continue
        xa=S[a]^val
        for b in range(a+1,m):
            if b==idx: continue
            if (xa&(S[b]^val))==0:
                e+=1
    # idx as a leg, apex j != idx, other leg b != idx, b>idx-as-leg handled by counting all b!=idx,j
    for j in range(m):
        if j==idx: continue
        q=S[j]
        xi=val^q
        for b in range(m):
            if b==idx or b==j: continue
            if (xi&(S[b]^q))==0:
                e+=1   # ordered leg pair (idx,b); each unordered pair counted twice over j-loop? no: apex fixed j, legs {idx,b} unordered counted once when we restrict...
    return e

# The leg-as-pair double counts: for apex j, pair {idx,b} should count once. Above loop counts
# it once per b (idx fixed as one leg, b the other) => once. Good. But the apex-block also can't
# overlap. So vertex_contrib is the exact number of violating unordered (apex,legpair) incidences
# touching idx. We use it only for relative delta of single-vertex replacement (consistent).

def replace_delta(S, idx, newval):
    old=vertex_contrib(S, idx, S[idx])
    new=vertex_contrib(S, idx, newval)
    return new-old

def run(n, k, seeds, iters, seedfile, out_prefix):
    rng=random.Random(12345)
    W=parse_witnesses(seedfile) if seedfile else {}
    seed_masks=W.get(n, [])
    full=(1<<n)
    best_global=None; best_e=10**9
    for s in range(seeds):
        rng.seed(1000*s+7)
        # init: seed from record set + random extras, or fully random
        S=list(seed_masks)
        random.Random(s).shuffle(S)
        if len(S)>k: S=S[:k]
        seen=set(S)
        while len(S)<k:
            v=rng.randrange(full)
            if v not in seen:
                seen.add(v); S.append(v)
        E=total_energy(S)
        T=2.0; cool=0.9997
        stagn=0; localbest=E
        for it in range(iters):
            if E==0: break
            # pick a vertex involved in many violations preferentially sometimes
            idx=rng.randrange(k)
            newval=rng.randrange(full)
            if newval in seen and newval!=S[idx]:
                continue
            d=replace_delta(S, idx, newval)
            if d<=0 or rng.random()<pow(2.718281828, -d/max(T,1e-9)):
                seen.discard(S[idx]); S[idx]=newval; seen.add(newval)
                E+=d
                if E<localbest:
                    localbest=E; stagn=0
                else:
                    stagn+=1
            T*=cool
            if stagn>4000:
                # kick: randomize a few vertices
                for _ in range(3):
                    i2=rng.randrange(k)
                    nv=rng.randrange(full)
                    if nv in seen: continue
                    seen.discard(S[i2]); S[i2]=nv; seen.add(nv)
                E=total_energy(S); stagn=0; T=max(T,0.8)
        E=total_energy(S)  # exact recount
        if E<best_e:
            best_e=E; best_global=list(S)
        if E==0:
            # write witness
            fn=f"{out_prefix}_n{n}_k{k}_seed{s}.txt"
            with open(fn,'w') as f:
                for v in S:
                    f.write(''.join('1' if (v>>i)&1 else '0' for i in range(n))+'\n')
            print(f"ZERO ENERGY n={n} k={k} seed={s} -> {fn}")
            return 0, fn, best_global
        print(f"  seed {s}: best E={E}", flush=True)
    # write best-so-far for record
    fn=f"{out_prefix}_n{n}_k{k}_BEST_e{best_e}.txt"
    with open(fn,'w') as f:
        for v in best_global:
            f.write(''.join('1' if (v>>i)&1 else '0' for i in range(n))+'\n')
    print(f"BEST n={n} k={k} energy={best_e} -> {fn}")
    return best_e, fn, best_global

if __name__=="__main__":
    n=int(sys.argv[1]); k=int(sys.argv[2])
    seeds=int(sys.argv[3]) if len(sys.argv)>3 else 20
    iters=int(sys.argv[4]) if len(sys.argv)>4 else 200000
    seedfile=sys.argv[5] if len(sys.argv)>5 else None
    out_prefix=sys.argv[6] if len(sys.argv)>6 else "wit"
    run(n,k,seeds,iters,seedfile,out_prefix)

File: code/acute_pilot/penalty_sa/sa.c
/* Fixed-cardinality penalty SA for OEIS A089676 acute sets in {0,1}^n.
 *
 * Energy = # unordered (apex j; leg pair {a,b}) with ((S[a]^S[j])&(S[b]^S[j]))==0.
 * Hold k fixed, drive energy->0 by single-vertex replacement (SA).
 * energy==0 => size-k acute set (verify externally with verify.py).
 *
 * Build:  cc -O3 -march=native -o sa sa.c
 * Run:    ./sa <n> <k> <seeds> <iters> <seed0/1> <rngbase> [out_prefix]
 *   seed0/1 = 1 to seed from record set hardcoded? No -> we read seeds from stdin
 *   We read optional seed masks (one decimal mask per line) from a file given as last arg.
 *
 * Actual CLI used by driver:
 *   ./sa n k seeds iters rngbase seedfile out_prefix
 */
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <math.h>
#include <stdint.h>

typedef uint32_t u32;
typedef uint64_t u64;

static u64 rs;
static inline u64 xrand(){ rs^=rs<<13; rs^=rs>>7; rs^=rs<<17; return rs; }
static inline double urand(){ return (xrand()>>11)*(1.0/9007199254740992.0); }

int n,k;
u32 mask_full;
u32 *S;        /* current vertices, size k */

/* contribution of position idx if its value is val: number of violating
 * (apex, legpair) incidences touching idx. Matches Python vertex_contrib. */
static long contrib(int idx, u32 val){
    long e=0;
    /* idx as apex */
    for(int a=0;a<k;a++){ if(a==idx) continue;
        u32 xa=S[a]^val;
        for(int b=a+1;b<k;b++){ if(b==idx) continue;
            if((xa&(S[b]^val))==0) e++;
        }
    }
    /* idx as a leg, apex j!=idx, other leg b!=idx,j */
    for(int j=0;j<k;j++){ if(j==idx) continue;
        u32 q=S[j]; u32 xi=val^q;
        for(int b=0;b<k;b++){ if(b==idx||b==j) continue;
            if((xi&(S[b]^q))==0) e++;
        }
    }
    return e;
}

static long total_energy(){
    long e=0;
    for(int j=0;j<k;j++){ u32 q=S[j];
        for(int a=0;a<k;a++){ if(a==j) continue; u32 xa=S[a]^q;
            for(int b=a+1;b<k;b++){ if(b==j) continue;
                if((xa&(S[b]^q))==0) e++;
            }
        }
    }
    return e;
}

static int contains(u32 v){ for(int i=0;i<k;i++) if(S[i]==v) return 1; return 0; }

int main(int argc,char**argv){
    if(argc<6){ fprintf(stderr,"usage: sa n k seeds iters rngbase seedfile [prefix]\n"); return 2; }
    n=atoi(argv[1]); k=atoi(argv[2]);
    int seeds=atoi(argv[3]); long iters=atol(argv[4]);
    u64 rngbase=strtoull(argv[5],0,10);
    const char*seedfile=(argc>6)?argv[6]:0;
    const char*prefix=(argc>7)?argv[7]:"wit";
    mask_full=(n>=32)?0xffffffffu:((1u<<n)-1);

    /* read seed masks */
    u32 *seedm=0; int nseed=0,cap=0;
    if(seedfile){ FILE*f=fopen(seedfile,"r");
        if(f){ char line[256];
            while(fgets(line,sizeof line,f)){ char*p=line; while(*p==' '||*p=='\t')p++;
                if(*p<'0'||*p>'9') continue; u32 v=(u32)strtoul(p,0,10);
                if(nseed==cap){ cap=cap?cap*2:64; seedm=realloc(seedm,cap*sizeof(u32)); }
                seedm[nseed++]=v;
            } fclose(f);
        }
    }

    S=malloc(k*sizeof(u32));
    long best_e=1L<<60; u32 *best=malloc(k*sizeof(u32));

    for(int s=0;s<seeds;s++){
        rs = rngbase*2654435761u + (u64)s*40503u + 0x9e3779b97f4a7c15ULL;
        for(int w=0;w<7;w++) xrand();
        /* init: shuffle seed masks, take first up-to-k, fill random distinct */
        int used=0;
        if(nseed>0){
            /* Fisher-Yates a copy */
            u32 *tmp=malloc(nseed*sizeof(u32)); memcpy(tmp,seedm,nseed*sizeof(u32));
            for(int i=nseed-1;i>0;i--){ int j=xrand()%(i+1); u32 t=tmp[i];tmp[i]=tmp[j];tmp[j]=t; }
            int take = nseed<k?nseed:k;
            for(int i=0;i<take;i++) S[used++]=tmp[i];
            free(tmp);
        }
        while(used<k){ u32 v=xrand()&mask_full; if(!contains(v)){ S[used++]=v; } }

        long E=total_energy();
        /* per-vertex involvement (violations touching i), for biased selection */
        long *inv=malloc(k*sizeof(long));
        for(int i=0;i<k;i++) inv[i]=contrib(i,S[i]);
        long bestE=E; u32 *bestS=malloc(k*sizeof(u32)); memcpy(bestS,S,k*sizeof(u32));
        double Thi=2.5, Tlo=0.05;
        double T=Thi; long stagn=0;
        /* geometric cool over a "cycle"; on stagnation, reheat (reanneal) */
        long cyclen = iters/8; if(cyclen<20000) cyclen=20000;
        double cool=pow(Tlo/Thi, 1.0/(double)cyclen);
        for(long it=0; it<iters && E>0; it++){
            /* biased: with prob 0.8 pick a vertex weighted by involvement */
            int idx;
            if((xrand()&3)!=0){
                long tot=0; for(int i=0;i<k;i++) tot+=inv[i]+1;
                long r=(long)(xrand()% (u64)tot); idx=0;
                while(idx<k-1){ r-=inv[idx]+1; if(r<0) break; idx++; }
            } else idx=xrand()%k;
            u32 nv=xrand()&mask_full;
            if(contains(nv)) continue;
            long oldc=inv[idx];
            long newc=contrib(idx,nv);
            long d=newc-oldc;
            if(d<=0 || urand()<exp(-(double)d/(T>1e-9?T:1e-9))){
                S[idx]=nv; E+=d;
                /* update involvement vector incrementally is complex; recompute touched.
                   Cheap correct approach: recompute all inv (O(k^2)) only occasionally;
                   here just set idx and mark neighbors dirty by full recompute every accept
                   of changed vertex -> too slow. Instead recompute inv for all (O(k^2)). */
                for(int i=0;i<k;i++) inv[i]=contrib(i,S[i]);
                if(E<bestE){ bestE=E; memcpy(bestS,S,k*sizeof(u32)); stagn=0; } else stagn++;
            } else stagn++;
            T*=cool; if(T<Tlo) T=Tlo;
            if(stagn>cyclen){
                /* reheat + restart from best with a perturbation */
                memcpy(S,bestS,k*sizeof(u32));
                for(int r=0;r<2;r++){ int i2=xrand()%k; u32 vv=xrand()&mask_full; if(!contains(vv)) S[i2]=vv; }
                E=total_energy();
                for(int i=0;i<k;i++) inv[i]=contrib(i,S[i]);
                stagn=0; T=Thi;
            }
        }
        E=total_energy();
        if(bestE<E){ memcpy(S,bestS,k*sizeof(u32)); E=bestE; }
        free(inv); free(bestS);
        if(E<best_e){ best_e=E; memcpy(best,S,k*sizeof(u32)); }
        if(E==0){
            char fn[512]; snprintf(fn,sizeof fn,"%s_n%d_k%d_seed%d.txt",prefix,n,k,s);
            FILE*f=fopen(fn,"w");
            for(int i=0;i<k;i++){ for(int b=0;b<n;b++) fputc(((S[i]>>b)&1)?'1':'0',f); fputc('\n',f); }
            fclose(f);
            printf("ZERO ENERGY n=%d k=%d seed=%d -> %s\n",n,k,s,fn);
            fflush(stdout);
            return 0;
        }
        printf("  seed %d: best E=%ld\n",s,E); fflush(stdout);
    }
    char fn[512]; snprintf(fn,sizeof fn,"%s_n%d_k%d_BEST_e%ld.txt",prefix,n,k,best_e);
    FILE*f=fopen(fn,"w");
    for(int i=0;i<k;i++){ for(int b=0;b<n;b++) fputc(((best[i]>>b)&1)?'1':'0',f); fputc('\n',f); }
    fclose(f);
    printf("BEST n=%d k=%d energy=%ld -> %s\n",n,k,best_e,fn); fflush(stdout);
    return 0;
}
